### PR TITLE
ast: introduce 'Datum'

### PR DESCRIPTION
**Pull request by coocood**
Datum is a value box intended to replace 'interface{}' for better performance and easier to use.

### CODE REVIEW

File: ast/datum.go [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+	}
+}
+
+// SetValue sets any kind of value.
+func (d *Datum) SetValue(val interface{}) {
+	switch x := val.(type) {
+	case nil:
+		d.SetNull()
+	case int64:
+		d.SetInt64(x)
```
Comment (shenli) [inline ast/datum.go]: How about x.(type) == int?

Comment (coocood) [inline ast/datum.go]: Normally, we only use int64, int only used in test code.
It's ok to treat `int` as `interface{}`

File: parser/parser.y [hunk 2]
```diff
@@ -1867,15 +1867,6 @@ Literal:
 |	floatLit
 |	intLit
 |	stringLit
-	{
-		tp := types.NewFieldType(mysql.TypeString)
-		l := yylex.(*lexer)
```
Comment (shenli) [inline parser/parser.y]: Miss charset info.

File: ast/datum.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+	KindInt64 int = iota + 1
+	KindUint64
+	KindFloat64
+	KindString
+	KindBytes
+	KindInterface
+)
+
+// Datum is a data box holds different kind of data.
+// It has better performance and easier to use than `interface{}`
```
Comment (zimulala) [inline ast/datum.go]: add . at the end of the line

File: parser/parser.y [hunk 4]
```diff
@@ -1869,12 +1869,11 @@ Literal:
 |	stringLit
 	{
 		tp := types.NewFieldType(mysql.TypeString)
-		l := yylex.(*lexer)
-		tp.Charset, tp.Collate = l.GetCharsetInfo()
-		$$ = &types.DataItem{
-			Type: tp,
-			Data: $1.(string),
-		}
+        l := yylex.(*lexer)
```
Comment (shenli) [inline parser/parser.y]: alignment

**Comment by shenli** [comment]
LGTM

**Comment by zimulala** [comment]
need some tests？

**Comment by zimulala** [comment]
LGTM

**Pull request merged**